MobiFlight/Joysticks/VKB/VKBDevice.cs

```diff
@@ -0,0 +1,192 @@
+        private readonly VKBLedContainer Lights = new VKBLedContainer();
+        private HidDeviceInputReceiver InputReceiver;
+        private readonly byte[] InputReportBuffer = new byte[64];
+        private readonly SortedList<byte, VKBEncoder> Encoders = new SortedList<byte, VKBEncoder>();
+        int lastSeqNo = -1;
+
+        public VKBDevice(SharpDX.DirectInput.Joystick joystick, JoystickDefinition definition) : base(joystick, definition)
+        {
+            Definition = definition;
+            if (Device == null)
+            {
+                Device = GetMatchingHidDevice(joystick);
+                if (Device == null) return;
+            }
+        }
+        public override void Connect(IntPtr handle)
+        {
+            base.Connect(handle);
+            if (Stream == null)
+            {
+                Stream = Device.Open();
+                Stream.ReadTimeout = System.Threading.Timeout.Infinite;
+            }
+
+            if (InputReceiver == null)
+            {
+                InputReceiver = new ReportDescriptor(VKBHidReport.Descriptor).CreateHidDeviceInputReceiver();
+                // We use our own descriptor since the one we get from Windows/HIDSharp has been altered and encoder data is missing
+                InputReceiver.Received += OnHidReportReceived;
+                InputReceiver.Start(Stream);
+            }
+        }
+        protected override void SendData(byte[] data)
+        {
+            // VKBLedContainer has its own handling that replaces RequiresOutputUpdate, so we just send.
+            // Don't try and send data if no outputs are defined.
+            if (Definition?.Outputs == null || Definition?.Outputs.Count == 0)
+            {
+                return;
+            }
+            Stream?.SetFeature(data);
+
+        }
+        protected override void EnumerateDevices()
+        {
+            base.EnumerateDevices();
+            SortedList<byte, JoystickDevice> EncoderDecList = new SortedList<byte, JoystickDevice>();
+            SortedList<byte, JoystickDevice> EncoderIncList = new SortedList<byte, JoystickDevice>();
+            if (Definition == null || Definition.Inputs == null) return;
+            foreach (var input in Definition.Inputs)
+            {
+                // The 1xxx range is limited to encoders. They are also not fed from DirectInput.
+                if (input.Id >= 1000 && input.Id < 2000 && input.Type == JoystickDeviceType.Button)
+                {
+                    byte encoderIndex = GetEncoderIndex(input);
+                    VKBEncoder.EncoderAction encoderAction = GetEncoderAction(input);
+                    if (encoderAction == VKBEncoder.EncoderAction.DEC)
+                    {
+                        EncoderDecList.Add(encoderIndex, new JoystickDevice { Name = $"Button {1000 + 10 * encoderIndex + (int)encoderAction}", Label = input.Label, Type = DeviceType.Button, JoystickDeviceType = input.Type });
```

Comment:
Can you add a comment that shows what the formatted display name looks like? It's a little tough to tell from the string interpolation.

Reply:
I have it in the other place encoders can be created (when they are found on the device but not in the definition), but I can reproduce it here.